mlir/lib/Dialect/Mesh/IR/MeshOps.cpp

```diff
@@ -129,6 +205,347 @@ MeshShardingAttr::verify(function_ref<InFlightDiagnostic()> emitError,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// collective communication ops
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+std::optional<DenseI16ArrayAttr>
+canonicalizeAxesSetAttribute(DenseI16ArrayAttr attr) {
```

Comment:
Why do you regard `mesh_axis` as a set? I think the duplication should not be allowed in verification.

Reply:
I was thinking to change its meaning. In the Mesh.md doc it is stated that the order of devices is the same as the order of  their multi-index in the mesh. This would make it a set.
We can change the meaning where the order of devices could be induced by the order in `mesh_axes`. For example if `mesh_axes = [2, 1]` then device `(i, j, k)` would precede device `(i, l, m)` if `k > m`. This would influence operations like all-gather where order device order is important.

Reply:
But it seems your reply cannot explain why we have to make `mesh_axes` support duplications like [2, 1, 2]?

Reply:
I wanted to use a set but there is no such attribute, so I went for an array. In math the set `{2, 1, 2}` is the same as `{2, 1}`. I can see why you want to forbid duplication as it may result from a programming error where the intention was something else.

Reply:
Yeah, I think not-accepting `{2, 1, 2}` could make the semantic more clear.

Reply:
Done.
I also change the meaning of `mesh_axes` to induce the order of devices in the device groups. I added documentation about this.